packages/@orbit/jsonapi/src/jsonapi-serializer.ts

```diff
@@ -185,30 +185,37 @@ export default class JSONAPISerializer {
     }
   }
 
-  deserializeDocument(document: JSONAPIDocument): DeserializedDocument {
+  deserializeDocument(document: JSONAPIDocument, primaryRecordData?: Record | Record[]): DeserializedDocument {
     let result: DeserializedDocument;
 
-    let data = document.data;
-    if (isArray(data)) {
-      result = {
-        data: (<Resource[]>data).map(this.deserializeResource, this)
-      };
-    } else if(data !== null) {
-      result = {
-        data: this.deserializeResource(<Resource>data)
-      };
+    let data;
+    if (isArray(document.data)) {
+      if (primaryRecordData !== undefined) {
+        data = (<Resource[]>document.data).map((entry, i) => {
+          return this.deserializeResource(entry, primaryRecordData[i]);
+        });
+      } else {
+        data = (<Resource[]>document.data).map((entry, i) => this.deserializeResource(entry));
+      }
+    } else if (document.data !== null) {
+      if (primaryRecordData !== undefined) {
+        data = this.deserializeResource(<Resource>document.data, (<Record>primaryRecordData));
+      } else {
+        data = this.deserializeResource(<Resource>document.data);
+      }
     } else {
-      result = { data: null };
+      data = null;
     }
+    result = { data };
 
     if (document.included) {
-      result.included = document.included.map(this.deserializeResource, this);
+      result.included = document.included.map(e => this.deserializeResource(e));
     }
 
     return result;
   }
 
-  deserializeResource(resource: Resource): Record {
+  deserializeResource(resource: Resource, primaryRecord?: Record): Record {
```

Comment:
Where in this function should we assign the id of the existing local record?

Reply:
Just give it precedence over any call to `this.schema.generateId(type);`?

Reply:
Yeah, essentially:

```
       if (resource.id) {
         keys = {
           [resourceKey]: resource.id
         };
 
         id = (primaryRecord && primaryRecord.id) ||
              this.keyMap.idFromKeys(type, keys) ||
              this.schema.generateId(type);
       } else {
         id = (primaryRecord && primaryRecord.id) || 
              this.schema.generateId(type);
       }
```